osu.Game/Screens/Select/PlaySongSelect.cs

```diff
@@ -30,6 +34,13 @@ public partial class PlaySongSelect : SongSelect
 
         public override bool AllowExternalScreenChange => true;
 
+        public override Func<BeatmapInfo, MenuItem>[] CustomMenuItems =>
+            new Func<BeatmapInfo, MenuItem>[]
+            {
+                b => new OsuMenuItem("Play", MenuItemType.Highlighted, () => FinaliseSelection(b)),
+                b => new OsuMenuItem(Resources.Localisation.Web.CommonStrings.ButtonsEdit, MenuItemType.Standard, () => Edit(b))
+            };
```

Comment:
This sort of least-path-of-resistance solution always brings me teeth pain when I see one now. Like sure, this works now, but what happens when N months the line someone wants to rearrange items at the *end* of the list? Or, say, put one in the middle?

There's probably about a fifty-fifty chance of this happening, and it has happened before, several times, in this repository, and has brought more teeth pain. I do not wish for it to happen again. A more proper direction for me would entail having each song select somehow define the _entirety_ of context menu items valid for the given variant, start-to-end. And sure, that may mean some duplication - but probably not *that* much if you structure this smart.

Sketch of imagined solution: Split a class for the menu, have it define all possible operations as methods, have one `CreateMenuItems()` method that returns all of the items in the proper order and calls the appropriate method for each one as the `Action`. Have it be created in `SongSelect` and have it overridable. Change returned implementation in online song select to achieve desired effect.

Reply:
I understand your concern and like your imagined solution. I shall change this code.

Reply:
Hmm. I'm not sure I share this concern. I see the menu items provided by implementation as only ever being the actions to proceed forwards (as generally limited by `FinaliseSelection` being the point of override for the non-standard song select implementations). I think going any further is likely over-engineering until we actually come across a use case.

I've cleaned this up to be a method and generally feel okay to me, @bdach could you take another look?

Reply:
Actually I agree with peppy

Reply:
Eh. If you can guarantee me that these methods will *only ever* return actions that make sense to be primary/topmost then sure I could live with this. However at this point I've a deep distaste for weird, overfitted, and underspecified virtual methods that essentially just end up being ordering implementation details that cause problems down the line. I've seen too many and don't wish to see more, and this looks like one for me if I squint at it enough (what is a "menu item for beatmap"? why does it not include collection management or deletion options, actually? are they not "menu items for beatmap"? I would at least xmldoc this method to explain the purpose better, or maybe name it `CreateNavigationalMenuItemsForBeatmap()`? `CreatePrimaryMenuItemsForBeatmap()`? or something)

Examples of what I mean elsewhere: [this](https://github.com/ppy/osu/pull/21497#discussion_r1064580078) (only proposed, but I think it's a bad idea), or [this](https://github.com/ppy/osu/blob/56556e0eca5f7b03224053d232f03cb6b0a07d56/osu.Game/Overlays/Rankings/Tables/UserBasedTable.cs#L76-L85) (crowning example of virtual badness).

My point is that I'd rather have this sort of implementation-specific stuff defined in a declarative manner in a single place, even at a cost of duplication, so that it doesn't require chasing through five files to figure out the final order of menu items in the damned thing. At least "create the entire context menu" for me is a clear and atomic behaviour at no risk of having weird bojangles done on it because `$screen` wants its items ordered thataway and not thisaway.

Reply:
Yeah I get that, but the requirements to make that happen is not as easy as having one method (as you mentioned it would need some separate component to actually create the common ones), which is the point where I feel it's over-engineered for what it's doing.

I think renaming to `CreateForwardNavigationMenuItemsForBeatmap` may be a good compromise?

Reply:
Works for me.